### PR TITLE
view functor generation

### LINKED ISSUE #142: Add a nice wrapper class to avoid implementing pipe-operator for custom views.
Please note, this is a design question/feature proposal.
Inspired by the nice tutorial about implementing your own ranges in the wiki, I was thinking that it would be great to take the burden off of the developers to always implement the pipe-operator for their custom views.
Especially, as it could be very confusing in the usage if stuff like `std::bind` together with placeholders is used to implement proper pipe-operator delegation for the respective view.
Hence, I came up with a CRTP pattern, that in general would allow us to implement a very handsome 
`my_view_fn` class where I don't need to think about the implementation of the pipe-operator and these kind of stuff.
Of course some stuff has to be dealt with, like proper const-fication, but in general I think that would be the way to go.
It might be possible, that I wasn't thinking of some corner cases so this should be discussed here, and we need to see if we can update the proposal or if we have to withdraw completely.

**[-note-begin-**
Of course if the view would have two constructors, while expecting 2 arguments and one of the constructors has a default parameter for the second argument, than there would be probably an ambiguous function call. Then again, the overloaded function call operator of `view_fn` could handle the default argument and thus making this design still applicable I think.
**-end-]**

But now enough talking. Lean back and enjoy! 😁 

```cpp

#include <vector>
#include <utility>
#include <memory>
#include <tuple>

namespace seqan3
{

// The factory that delegates the pipe to a custom call using the lhs of a |-operator
// implemented as a CRTP class.
template <typename derived_view_fn_t>
struct view_factory
{
    // A reference to the derived view_fn object.
    derived_view_fn_t & view_fn;

    // Construction.
    view_factory(derived_view_fn_t & derived) : view_fn(derived)
    {}

    // A special functor that gives us a named type for which we can overload the
    // pipe-operator. NOTE: I tried to use a lambda instead, but obviously, what should
    // the signature of the pipe operator look like? It would get two template parameters,
    // making it applicable for everything.
    template <typename... arg_ts>
    struct view_factory_fn
    {
        // Get access to the underlying view_fn.
        derived_view_fn_t & view_fn;

        // Store the variadic values somewhere.
        std::tuple<arg_ts...> member;

        // constructor.
        view_factory_fn(derived_view_fn_t & fn, arg_ts && ...args) :
            view_fn(fn),
            member{std::forward<arg_ts>(args)...}
        {}

        // helper class to unpack the tuple.
        template <typename rng_t, size_t... Is>
        auto explode(rng_t && rng, std::index_sequence<Is...> const &)
        {
            return view_fn(std::forward<rng_t>(rng), std::forward<arg_ts>(std::get<Is>(member))...);
        }

        // the operator called by the pipe-operator.
        auto operator()(auto && lhs_rng)
        {
            return explode(std::forward<decltype(lhs_rng)>(lhs_rng), std::make_index_sequence<sizeof...(arg_ts)>{});
        };
    };

    // the operator called by the derived class to get a proxy for calling the pipe-operator later on
    // without using the weird interface with std::bind.
    template <typename... arg_ts>
    auto operator()(arg_ts && ...args)
    {
        return view_factory_fn<arg_ts...>{view_fn, std::forward<arg_ts>(args)...};
    }

    // the actual magic pipe-operator with getting view_factory_fn on the rhs,
    // so that it does not get to be called for every type.
    template <typename rng_t, typename... arg_ts>
    friend auto operator|(rng_t && lhs, view_factory_fn<arg_ts...> && callable)
    {
        return callable(std::forward<rng_t>(lhs));
    }
};

// Some test view implementation that does not do a lot.
template <typename rng_t = void>
class my_view_impl
{
protected:

    std::shared_ptr<std::decay_t<rng_t>> data{};
    unsigned member1{};
public:
    my_view_impl() = default;

    my_view_impl(unsigned const m1) noexcept : member1(m1)
    {}

    my_view_impl(rng_t && rng, unsigned const m1) noexcept : data{std::make_shared<std::decay_t<rng_t>>(rng)}, member1{m1}
    {}

    unsigned get() const noexcept
    {
        return member1;
    }
};

// That would be an example .._fn implementation using the CRTP pattern,
// to get almost automatically access to the pipe notation.
struct my_view_fn : protected view_factory<my_view_fn>
{
    // the base class.
    using factory_t = view_factory<my_view_fn>;

    // the constructor.
    my_view_fn() : factory_t(*this)
    {};

    // Just tell the compiler there is a suitable operator() overload that apparently
    // does all the magic for us.
    using factory_t::operator();

    // Now specify what should happen if the final operator is called.
    // This allows us to use the functional call design or the pipe notation.
    template <typename rng_t>
    auto operator()(rng_t && rng, unsigned const p1)
    {
        return my_view_impl<rng_t>(std::forward<rng_t>(rng), p1);
    }
};

namespace view
{
// Well the code above is yet to be const-ified.
my_view_fn /*const*/ my_view;
}
}

using namespace seqan3;

// Some general use cases, that should work.
int main()
{
    { // use case 1: functional;
        std::vector<int> vec{{1, 2, 3, 4, 5}};
        auto v = view::my_view(vec, 4u);
        std::cout << v.get()  << "\n"; // should print 4
    }

    {  // use case 2: pipe version.
        std::vector<int> vec{{1, 2, 3, 4, 5}};
        auto v2 = vec | view::my_view(4u) | view::my_view(3u);
        std::cout << v.get()  << "\n"; // should print 3
    }
}
```

### PR DESCRIPTION
**Pull request by h-2**
fixes #142

### CODE REVIEW

File: include/seqan3/range/view/detail.hpp [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+     * ```
+     */
+    template <typename ...arg_types>
+    auxiliary_functor_t<std::add_rvalue_reference_t<arg_types>...> operator()(arg_types && ...args) const
+    {
+        // adding rvalue reference is important, because temporaries+&& make arg_type
+        // be e.g. "int" and not "int &&", but we really want the tuple in the functor
+        // to always only contain lvalue or rvalue references
+        // [adding rvalue reference to lvalue reference is NOP which we want]
+        return {std::add_rvalue_reference_t<arg_types>(args)...};
```
Comment (h-2) [inline include/seqan3/range/view/detail.hpp]: I found this last-minute, before it generated dangling stack references if the argument type inside the view was `&`.
:sweat_smile:

Comment (h-2) [reply include/seqan3/range/view/detail.hpp]: hm, or maybe I did fix it :thinking:

**Comment by h-2** [comment]
ok, done from my POV

File: test/range/view/declare_view_functor_type_test.cpp [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+#include <gtest/gtest.h>
+
+#include <seqan3/range/view/detail.hpp>
+
+using namespace seqan3;
+
+/*************++ without args ************************/
+
+// not actually a view
+template <typename urng_t>
```
Comment (marehr) [inline test/range/view/declare_view_functor_type_test.cpp]: sry for the stupid question, but what is a `urange`? what does the `u` stand for?

Comment (h-2) [reply test/range/view/declare_view_functor_type_test.cpp]: TL;DR: u stands for **underyling** range. It is the range object this views operates on.

Since view functors operate on another range and return a new range (more precisely a view), I used to call the types the input range (`irng_t irange`) and the output range. However this is confusing, because input range and output range are actually concepts that describe certain characteristics of the range, not whether they are returned from a functor or put into one...
So the range returned is actually also a seqan3::input_range_concept and not a seqan3::output_range_concept. Seeing that this was confusing I already renamed the "output range" to "returned range" in our documentation.

I now discussed with @rrahn that we also rename the input to "**underlying** range". This is a TODO for the existing views, but in this PR and #154, well as our Tutorial already adapt this naming.

**Comment by marehr** [comment]
Ah okay. underlying_range_t is to long? Rng is for me a random number generator. But I had a hinge that you mean ranges. I imagine that using range_t is to under specified. If we stick to urng_t this should be explained within each doc for the template argument. I found it only in one. 

Thank you for clarification 

Am 3. März 2018 12:27:07 MEZ schrieb Hannes Hauswedell <notifications@github.com>:
>h-2 commented on this pull request.
>
>
>
>> +//
>+//
>==========================================================================
>+
>+#include <iostream>
>+#include <memory>
>+
>+#include <gtest/gtest.h>
>+
>+#include <seqan3/range/view/detail.hpp>
>+
>+using namespace seqan3;
>+
>+/*************++ without args ************************/
>+
>+// not actually a view
>+template <typename urng_t>
>
>TL;DR: u stands for **underyling** range. It is the range object this
>views operates on.
>
>Since view functors operate on another range and return a new range
>(more precisely a view), I used to call the types the input range
>(`irng_t irange`) and the output range. However this is confusing,
>because input range and output range are actually concepts that
>describe certain characteristics of the range, not whether they are
>returned from a functor or put into one...
>So the range returned is actually also a seqan3::input_range_concept
>and not a seqan3::output_range_concept. Seeing that this was confusing
>I already renamed the "output range" to "returned range" in our
>documentation.
>
>I now discussed with @rrahn that we also rename the input to
>"**underlying** range". This is a TODO for the existing views, but in
>this PR and #154, well as our Tutorial already adapt this naming.
> 
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/seqan/seqan3/pull/153#discussion_r172014362

-- 
Diese Nachricht wurde von meinem Android-Gerät mit K-9 Mail gesendet.

File: include/seqan3/range/view/detail.hpp [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+     */
+    template <typename ...arg_types>
+    struct auxiliary_functor_t
+    {
+        //!\brief Store the arguments
+        std::tuple<arg_types...> arguments;
+
+        //!\brief Helper function to unpack the tuple.
+        template <typename urng_t, size_t... Is>
+        auto explode(urng_t && urange, std::index_sequence<Is...> const &)
```
Comment (rrahn) [inline include/seqan3/range/view/detail.hpp]: Just minor thing. This function should be scoped in a protected region. 
Not really an exposable interface. 
The same holds for the `arguments` member.

**Comment by h-2** [comment]
@rrahn I think I addressed all your concerns. For the arguments member of the intermediate functor type I have renamed it `_arguments`, because make a data member non-public prevents aggregate initialisation leading to half a page of constructors having to be defined which would make the whole thing more difficult to read, I think.

**Comment by rrahn** [comment]
lgtm
